File: src/numerical_pdes/boundary_conditions.py
from abc import ABC, abstractmethod
from numpy.typing import NDArray


class BoundaryCondition(ABC):
    @abstractmethod
    def apply(self, u: NDArray, n_ghost_cells: int) -> None:
        """Fill volume ghost cells in-place."""
        ...

    @abstractmethod
    def apply_interface(self, u_right: NDArray, u_left: NDArray, n_cells: int) -> None:
        """Fill interface boundary values in-place after reconstruction."""
        ...


class PeriodicBC(BoundaryCondition):
    def apply(self, u: NDArray, n_ghost_cells: int) -> None:
        n_interior = u.shape[0] - 2 * n_ghost_cells
        for i in range(n_ghost_cells):
            u[i, :] = u[i + n_interior, :]
            u[-1 - i, :] = u[-1 - i - n_interior, :]

    def apply_interface(self, u_right: NDArray, u_left: NDArray, n_cells: int) -> None:
        u_left[0] = u_left[n_cells]
        u_right[n_cells] = u_right[0]


class WallBC(BoundaryCondition):
    def apply(self, u: NDArray, n_ghost_cells: int) -> None:
        for i in range(n_ghost_cells):
            u[i, :] = u[2 * n_ghost_cells - 1 - i, :]
            u[-1, :] = u[-(2 * n_ghost_cells) + i, :]

    def apply_interface(self, u_right: NDArray, u_left: NDArray, n_cells: int) -> None:
        u_left[0] = u_right[0]
        u_right[n_cells] = u_left[n_cells]

File: src/numerical_pdes/csrc/boundary_conditions.cpp
#include "boundary_conditions.h"


void PeriodicBC::Apply(torch::Tensor& u, int n_ghost_cells) const {
  int n_cells = u.size(1) - 2 * n_ghost_cells;
  for (int i=0; i < n_ghost_cells; i++) {
    u.select(/*dim=*/1, i).copy_(u.select(/*dim=*/1, i + n_cells));
    u.select(/*dim=*/1, n_ghost_cells + n_cells + i).copy_(u.select(/*dim=*/1, n_ghost_cells + i));
  }
}


void PeriodicBC::ApplyInterface(torch::Tensor& u_right, torch::Tensor& u_left, int n_cells) const {
  u_left.select(/*dim=*/1, 0).copy_(u_left.select(/*dim=*/1, n_cells));
  u_right.select(/*dim=*/1, n_cells).copy_(u_right.select(/*dim=*/1, 0));
}


void WallBC::Apply(torch::Tensor& u, int n_ghost_cells) const {
  int n_cells = u.size(1) - 2 * n_ghost_cells;
  for (int i=0; i < n_ghost_cells; i++) {
    u.select(/*dim=*/1, i).copy_(u.select(/*dim=*/1, 2 * n_ghost_cells - 1 - i));
    u.select(/*dim=*/1, n_ghost_cells + n_cells + i).copy_(u.select(/*dim=*/1, n_ghost_cells + n_cells - 1 - i));
  }
}


void WallBC::ApplyInterface(torch::Tensor& u_right, torch::Tensor& u_left, int n_cells) const {
  u_left.select(/*dim=*/1, 0).copy_(u_right.select(/*dim=*/1, 0));
  u_right.select(/*dim=*/1, n_cells).copy_(u_left.select(/*dim=*/1, n_cells));
}
